src/psyclone/domain/lfric/kernel_arg_order.py

```diff
@@ -0,0 +1,1063 @@
+        name = self._cma_operator_name(cma_operator)
+        self.arg_info.append(f"cbanded_map_{function_space_name}_{name}")
+        self._arg_index += 1
+
+    def indirection_dofmap(self, function_space, cma_operator):
+        '''Adds an indirection dofmap for the provided function space and cma
+        operator when there is an apply cma kernel.
+
+        Include the indirection map for the 'to' function space of the
+        supplied CMA operator. This is a rank-1 integer array of kind
+        i_def and intent in with extent nrow. Its default name is
+        'cma_indirection_map_'<function_space>'_'<operator_name>.
+
+        If the from-space of the operator is not the same as the
+        to-space then include the indirection map for the 'from'
+        function space of the CMA operator. This is a rank-1 integer
+        array of kind i_def and intent in with extent ncol. Its
+        default name is
+        'cma_indirection_map_'<function_space>'_'<operator_name>.
+
+        Note, this method will not be called for the from space if the
+        to and from function spaces are the same so there is no need
+        to explicitly check.
+
+        :param str function_space: the function space for this \
+            indirection dofmap.
+        :param cma_operator: the cma operator metadata associated with \
+            this indirection dofmap.
+        :type cma_operator: :py:class:`psyclone.domain.lfric.kernel.\
+            ColumnwiseOperatorArgMetadata`
+
+        '''
+        function_space_name = self._function_space_name(function_space)
+        name = self._cma_operator_name(cma_operator)
+        self.arg_info.append(
+            f"cma_indirection_map_{function_space_name}_{name}")
+        self._arg_index += 1
+
+    def _generate(self, metadata):
+
+        '''Specifies which arguments appear in an argument list and their
+        ordering. Calls methods for each type of argument that can be
+        specialised by a child class for its particular need.
+
+        :param metadata: the LFRic kernel metadata.
+        :type metadata: \
+            py:class:`psyclone.domain.lfric.kernel.LFRicKernelMetadata`
+
+        :raises GenerationError: if the kernel arguments break the \
+                                 rules for the LFRic API.
+
+        '''
+        # All operator types require the cell index to be provided
+        if metadata.meta_args_get(
+                [OperatorArgMetadata, ColumnwiseOperatorArgMetadata]):
+            self.cell_position()
+
+        # Pass the number of layers in the mesh unless this kernel is
+        # applying a CMA operator or doing a CMA matrix-matrix calculation
+        if metadata.kernel_type not in ["cma-apply", "cma-matrix-matrix"]:
+            self.mesh_height()
+
+        # Pass the number of cells in the mesh if this kernel has a
+        # LMA operator argument
+        # TODO this code should replace the code that currently includes
```

Comment:
I've turned this into issue #2074 as it has been like this since the earliest versions of PSyclone. As you say we can discuss this with @TeranIvy but I suggest it is done separately.